component/loki/source/kubernetes/kubernetes.go

```diff
@@ -168,28 +186,43 @@ func (c *Component) Update(args component.Arguments) error {
 		// No-op: manager already exists and options didn't change.
 	}
 
-	// Convert input targets into targets to give to tailer.
-	targets := make([]*kubetail.Target, 0, len(newArgs.Targets))
+	c.resyncTargets(newArgs.Targets)
+	c.args = newArgs
+	return nil
+}
+
+func (c *Component) resyncTargets(targets []discovery.Target) {
+	distTargets := discovery.NewDistributedTargets(c.args.Clustering.Enabled, c.cluster, targets)
+	targets = distTargets.Get()
 
-	for _, inTarget := range newArgs.Targets {
-		lset := inTarget.Labels()
+	tailTargets := make([]*kubetail.Target, 0, len(targets))
+	for _, target := range targets {
+		lset := target.Labels()
 		processed, err := kubetail.PrepareLabels(lset, c.opts.ID)
 		if err != nil {
 			// TODO(rfratto): should this set the health of the component?
 			level.Error(c.log).Log("msg", "failed to process input target", "target", lset.String(), "err", err)
 			continue
 		}
-		targets = append(targets, kubetail.NewTarget(lset, processed))
+		tailTargets = append(tailTargets, kubetail.NewTarget(lset, processed))
 	}
 
 	// This will never fail because it only fails if the context gets canceled.
```

Comment:
Since this is called while holding the component mutex, I'd say yes, we should have a timeout. Even if we can only log it.

Reply:
I'm nervous about that; maybe it should be detached from the mutex instead. 

If it times out and fails, that means that the targets won't ever converge unless there's another discovery update or the cluster updates again. While discovery updates are common for us, it might not be common in more stable environments. 

So, the risk of a timeout causing targets to never converge is worrying me. WDYT?